File: docs/stack/code/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Node {
    int data;
    struct Node *next;
} Node;

typedef struct {
    Node *top;
    int size;
} Stack;

Stack create_stack() {
    Stack s;
    s.top = NULL;
    s.size = 0;
    return s;
}

void push(Stack *s, int data) {
    Node *node = (Node *)malloc(sizeof(Node));
    node->data = data;

    node->next = s->top;
    s->top = node;

    s->size++;
}

int pop(Stack *s) {
    if (s->top == NULL) {
        printf("Stack is empty\n");
        return -1;
    }
    Node *temp = s->top;
    int data = temp->data;

    s->top = s->top->next;
    s->size--;

    free(temp);

    return data;
}

int peek(Stack *s) {
    if (s->top == NULL) {
        printf("Stack is empty\n");
        return -1;
    }

    return s->top->data;
}

int is_empty(Stack *s) {
    return s->size == 0;
}

typedef struct CharNode {
    char data;
    struct CharNode *next;
} CharNode;

typedef struct {
    CharNode *top;
    int size;
} CharStack;

CharStack create_char_stack() {
    CharStack s;
    s.top = NULL;
    s.size = 0;
    return s;
}

void char_push(CharStack *s, char data) {
    CharNode *node = (CharNode *)malloc(sizeof(CharNode));
    node->data = data;
    node->next = s->top;
    s->top = node;
    s->size++;
}

char char_pop(CharStack *s) {
    CharNode *temp = s->top;
    char data = temp->data;
    s->top = s->top->next;
    free(temp);
    s->size--;
    return data;
}

int char_is_empty(CharStack *s) {
    return s->size == 0;
}

int is_balanced(char *str) {
    CharStack s = create_char_stack();

    int len = strlen(str);
    for (int i = 0; i < len; i++) {
        char c = str[i];

        if (c == '(' || c == '[' || c == '{') {
            char_push(&s, c);
        }

        else if (c == ')' || c == ']' || c == '}') {
            if (char_is_empty(&s)) {
                return 0;
            }

            char top = char_pop(&s);

            if ((c == ')' && top != '(') ||
                (c == ']' && top != '[') ||
                (c == '}' && top != '{')) {
                return 0;
            }
        }
    }

    return char_is_empty(&s);
}

typedef struct {
    Stack stack;
    Stack min_stack;
} MinStack;

MinStack create_min_stack() {
    MinStack ms;
    ms.stack = create_stack();
    ms.min_stack = create_stack();
    return ms;
}

void min_push(MinStack *ms, int data) {
    push(&ms->stack, data);

    if (is_empty(&ms->min_stack) || data <= peek(&ms->min_stack)) {
        push(&ms->min_stack, data);
    }
}

int min_pop(MinStack *ms) {
    int data = pop(&ms->stack);

    if (data == peek(&ms->min_stack)) {
        pop(&ms->min_stack);
    }

    return data;
}

int get_min(MinStack *ms) {
    if (is_empty(&ms->min_stack)) return -1;

    return peek(&ms->min_stack);
}

typedef struct {
    Stack in_stack;
    Stack out_stack;
} TwoStackQueue;

TwoStackQueue create_queue() {
    TwoStackQueue q;
    q.in_stack = create_stack();
    q.out_stack = create_stack();
    return q;
}

// Add an element to the back of the queue
void enqueue(TwoStackQueue *q, int data) {
    // Always push onto in_stack
    push(&q->in_stack, data);
}

// Remove and return the front element of the queue
int dequeue(TwoStackQueue *q) {
    // If out_stack is empty, pour in_stack into out_stack
    if (is_empty(&q->out_stack)) {
        while (!is_empty(&q->in_stack)) {
            // Pop from in_stack and push onto out_stack
            push(&q->out_stack, pop(&q->in_stack));
        }
    }

    // Pop from out_stack
    return pop(&q->out_stack);
}

int queue_peek(TwoStackQueue *q) {
    // If out_stack is empty, pour in_stack into out_stack
    if (is_empty(&q->out_stack)) {
        while (!is_empty(&q->in_stack)) {
            push(&q->out_stack, pop(&q->in_stack));
        }
    }

    // Peek from out_stack
    return peek(&q->out_stack);
}


int main() {
    Stack s = create_stack();
    push(&s, 10);       // TOP -> [10] -> NULL
    push(&s, 20);       // TOP -> [20] -> [10] -> NULL
    push(&s, 30);       // TOP -> [30] -> [20] -> [10] -> NULL

    int a = pop(&s);    // a = 30, TOP -> [20] -> [10] -> NULL
    int b = pop(&s);    // b = 20, TOP -> [10] -> NULL
    int c = pop(&s);    // c = 10, TOP -> NULL
    int d = pop(&s);    // d = -1, stack is empty

    printf("%d %d %d %d\n", a, b, c, d);

    is_empty(&s);       // 1 (true)

    push(&s, 10);       // TOP -> [10] -> NULL
    push(&s, 20);       // TOP -> [20] -> [10] -> NULL
    push(&s, 30);       // TOP -> [30] -> [20] -> [10] -> NULL

    int e = peek(&s);       // e = 30, stack unchanged
    int f = peek(&s);       // f = 30, stack unchanged

    printf("%d %d\n", e, f);

    is_empty(&s);       // 0 (false)

    printf("%d\n", is_balanced("([{}])"));  // 1 (true)
    printf("%d\n", is_balanced("([)]"));    // 0 (false)
    printf("%d\n", is_balanced("((("));     // 0 (false)

    MinStack ms = create_min_stack();
    min_push(&ms, 3);   // stack: [3],          min: [3]
    min_push(&ms, 5);   // stack: [3, 5],       min: [3]
    min_push(&ms, 2);   // stack: [3, 5, 2],    min: [3, 2]
    min_push(&ms, 1);   // stack: [3, 5, 2, 1], min: [3, 2, 1]

    printf("%d\n", get_min(&ms));   // 1

    min_pop(&ms);                   // stack: [3, 5, 2],    min: [3, 2]
    printf("%d\n", get_min(&ms));   // 2

    min_pop(&ms);                   // stack: [3, 5],       min: [3]
    printf("%d\n", get_min(&ms));   // 3

    // Usage
    TwoStackQueue q = create_queue();
    enqueue(&q, 1);     // [1]
    enqueue(&q, 2);     // [1, 2]
    enqueue(&q, 3);     // [1, 2, 3]

    printf("%d\n", dequeue(&q));    // 1
    printf("%d\n", dequeue(&q));    // 2

    enqueue(&q, 4);     // [3, 4]

    printf("%d\n", dequeue(&q));    // 3
    printf("%d\n", dequeue(&q));    // 4

    return 0;
}

File: docs/stack/code/stack.rs
struct Node<T> {
    data: T,
    next: Option<Box<Node<T>>>,
}

struct Stack<T> {
    top: Option<Box<Node<T>>>,
    size: usize,
}

impl<T> Stack<T> {
    fn new() -> Stack<T> {
        Stack { top: None, size: 0 }
    }

    fn push(&mut self, data: T) {
        let node = Box::new(Node {
            data,
            next: self.top.take(),
        });
        self.top = Some(node);
        self.size += 1;
    }

    fn pop(&mut self) -> Option<T> {
        let node = self.top.take()?;
        self.top = node.next;
        self.size -= 1;
        Some(node.data)
    }

    fn peek(&self) -> Option<&T> {
        self.top.as_ref().map(|node| &node.data)
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }
}

fn is_balanced(str: &str) -> bool {
    let mut s: Stack<char> = Stack::new();

    for c in str.chars() {
        match c {
            '(' | '[' | '{' => s.push(c),
            ')' | ']' | '}' => match s.pop() {
                None => return false,
                Some(top) => {
                    if (c == ')' && top != '(')
                        || (c == ']' && top != '[')
                        || (c == '}' && top != '{')
                    {
                        return false;
                    }
                }
            },
            _ => {}
        }
    }

    s.is_empty()
}

struct MinStack {
    stack: Stack<i32>,
    min_stack: Stack<i32>,
}

impl MinStack {
    fn new() -> MinStack {
        MinStack {
            stack: Stack::<i32>::new(),
            min_stack: Stack::<i32>::new(),
        }
    }

    fn push(&mut self, data: i32) {
        self.stack.push(data);

        if self.min_stack.is_empty() || data <= *self.min_stack.peek().unwrap() {
            self.min_stack.push(data);
        }
    }

    fn pop(&mut self) -> Option<i32> {
        let data = self.stack.pop()?;

        if Some(&data) == self.min_stack.peek() {
            self.min_stack.pop();
        }

        Some(data)
    }

    fn get_min(&self) -> Option<&i32> {
        self.min_stack.peek()
    }
}

struct TwoStackQueue {
    in_stack: Stack<i32>,
    out_stack: Stack<i32>,
}

impl TwoStackQueue {
    fn new() -> TwoStackQueue {
        TwoStackQueue {
            in_stack: Stack::new(),
            out_stack: Stack::new(),
        }
    }

    // Add an element to the back of the queue
    fn enqueue(&mut self, data: i32) {
        // Always push onto in_stack
        self.in_stack.push(data);
    }

    // Remove and return the front element of the queue
    fn dequeue(&mut self) -> Option<i32> {
        // If out_stack is empty, pour in_stack into out_stack
        if self.out_stack.is_empty() {
            while let Some(data) = self.in_stack.pop() {
                // Pop from in_stack and push onto out_stack
                self.out_stack.push(data);
            }
        }

        // Pop from out_stack
        self.out_stack.pop()
    }

    fn peek(&mut self) -> Option<&i32> {
        // If out_stack is empty, pour in_stack into out_stack
        if self.out_stack.is_empty() {
            while let Some(data) = self.in_stack.pop() {
                self.out_stack.push(data);
            }
        }

        // Peek from out_stack
        self.out_stack.peek()
    }
}

fn main() {
    let mut s = Stack::<i32>::new();

    s.push(10); // TOP -> [10] -> NONE
    s.push(20); // TOP -> [20] -> [10] -> NONE
    s.push(30); // TOP -> [30] -> [20] -> [10] -> NONE

    let a = s.pop(); // a = Some(30), TOP -> [20] -> [10] -> NONE
    let b = s.pop(); // b = Some(20), TOP -> [10] -> NONE
    let c = s.pop(); // c = Some(10), TOP -> NONE
    let d = s.pop(); // d = NONE, stack is empty

    println!("Popped values: {:?}, {:?}, {:?}, {:?}", a, b, c, d);

    s.is_empty(); // true

    if let Some(value) = s.pop() {
        println!("Popped value: {}", value);
    }

    match s.pop() {
        Some(value) => println!("Popped value: {}", value),
        None => println!("Stack is empty"),
    }

    s.push(10); // TOP -> [10] -> NONE
    s.push(20); // TOP -> [20] -> [10] -> NONE
    s.push(30); // TOP -> [30] -> [20] -> [10] -> NONE

    let e = s.peek(); // e = Some(30), TOP -> [30] -> [20] -> [10] -> NONE
    let f = s.peek(); // f = Some(30), TOP -> [30] -> [20] -> [10] -> NONE

    println!("Peeked values: {:?}, {:?}", e, f);

    s.is_empty(); // false

    println!("{}", is_balanced("([{}])")); // true
    println!("{}", is_balanced("([)]")); // false
    println!("{}", is_balanced("(((")); // false

    let mut ms = MinStack::new();
    ms.push(3); // stack: [3],          min: [3]
    ms.push(5); // stack: [3, 5],       min: [3]
    ms.push(2); // stack: [3, 5, 2],    min: [3, 2]
    ms.push(1); // stack: [3, 5, 2, 1], min: [3, 2, 1]

    println!("{:?}", ms.get_min()); // Some(1)

    ms.pop(); // stack: [3, 5, 2],    min: [3, 2]
    println!("{:?}", ms.get_min()); // Some(2)

    ms.pop(); // stack: [3, 5],       min: [3]
    println!("{:?}", ms.get_min()); // Some(3)

    // Usage
    let mut q = TwoStackQueue::new();
    q.enqueue(1); // [1]
    q.enqueue(2); // [1, 2]
    q.enqueue(3); // [1, 2, 3]

    println!("{:?}", q.dequeue()); // Some(1)
    println!("{:?}", q.dequeue()); // Some(2)

    q.enqueue(4); // [3, 4]

    println!("{:?}", q.dequeue()); // Some(3)
    println!("{:?}", q.dequeue()); // Some(4)
}
